### PR TITLE
Rename all non-snake_case types.

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do
This PR converts all non-snake_case type names to an appropriate alternative. It provides a commensurate UpdatePaths script and applies it. It provides a SQL script for the production feedback table. It adds a check to check_grep2 to prohibit such naming in the future. It adds a schema migration Python script for updating the feedback table (doing player and json_datum_saves now in a sec)

This refactoring was performed with the following procedure:

### Apply and store a renaming heuristic to offending type names

This was done with the script [`make_snake_case_json.py`][make]. The result was the JSON file [`snake_case_type_remap.json`][json]. Each element in the JSON file was a dictionary with "original" being the original path, and "replace" being the converted path.

### Clean up sub-optimal renames

Some of these replacements weren't great so I manually added an "override" key to those renames, so they wouldn't get clobbered by running the `make_snake_case.json` script again. For example:

```json
    // ...
	{
		"original": "/obj/item/clothing/head/HoS",
		"replace": "/obj/item/clothing/head/ho_s",
		"override": "/obj/item/clothing/head/hos"
	},
    // ...
```

### Generate the UpdatePaths script from the renames

[`generate_snake_case_updatepaths.py`][up] is a straightforward script which generated the UpdatePaths script. The results were spot-checked for correctness.

### Perform a search and replace on code

[`perform_snake_case_search_replace.py`][code] was used to perform the actual search and replace on the codebase, ensuring that parameter names and the like were all properly affected. The results were spot-checked and it was ensured the code compiled.

### Use the JSON file as input for a SQL migration.

The new schema migration script 60-61.py runs over all feedback rows and updates them based on the contents of the JSON renaming information. It is pretty fast:

```
PS D:\ExternalRepos\third_party\Paradise> python .\SQL\updates\60-61.py localhost para_migrations goaway parastats_prod_snake_case
[2024-11-02 23:23:40.101977] Connected to parastats_prod_snake_case
[2024-11-02 23:23:40.102977] Running query
[2024-11-02 23:23:40.103978] Fetching all data
[2024-11-02 23:23:41.318861] Processing 681114 rows...
[2024-11-02 23:23:59.895230] Running 2134 SQL updates
[2024-11-02 23:24:00.104232] Committing data
[2024-11-02 23:24:00.111231] Script done
```

## Why It's Good For The Game
~~ORDNUNG MUSS SEIN~~ Consistency in naming schemes is good because 1: it makes the code style more consistent, meaning less cognitive load for contributors, 2: it means not having to worry/think about case-sensitive regexes/searches looking for types, and 3: it means admins don't have to bang their heads against finding the right name for objects in the game menu.

## Testing
See above. The scripts and JSON file can be checked for correctness, and the results of the entire process recreated if necessary.

<hr>

### Declaration

- [X] I confirm that I either do not require [pre-approval](https://github.com/ParadiseSS13/Paradise/blob/master/docs/CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
<hr>

## Changelog
NPFC

[up]: https://gist.github.com/warriorstar-orion/d5c6aa90ad56d81f26813aacc6ec999c#file-generate_snake_case_updatepaths-py
[make]: https://gist.github.com/warriorstar-orion/d5c6aa90ad56d81f26813aacc6ec999c#file-make_snake_case_json-py
[json]: https://gist.github.com/warriorstar-orion/d5c6aa90ad56d81f26813aacc6ec999c#file-snake_case_type_remap-json
[code]: https://gist.github.com/warriorstar-orion/d5c6aa90ad56d81f26813aacc6ec999c#file-perform_snake_case_search_replace-py
[sql]: https://gist.github.com/warriorstar-orion/d5c6aa90ad56d81f26813aacc6ec999c#file-generate_snake_case_sql_script-py

### CODE REVIEW

**Comment by AffectedArc07** [comment]
@warriorstar-orion 

> I will try and keep the SQL script updated as time goes on but there will inevitably be a lag between what's in prod and what rows the SQL script touches. It should be straightforward to run the SQL generating script on the handful of rounds that were missed between submission and database change.

If before merge we halt new merges I can run the script and give you the repo output. Or you can make it like the other SQL update files that end in `.py` and have it apply as a migration in that fashion.

File: SQL/updates/60-61.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        host=args.address, user=args.username, passwd=args.password, db=args.database
+    )
+    cursor = db.cursor()
+    log(f"Connected to {args.database}")
+
+    # want these ordered by length from longest to shortest so shorter replacements
+    # don't replace pieces of larger replacements
+    renames = sorted(
+        json.load(
+            open(Path(__file__).parent / "migration_data/snake_case_type_remap.json")
```
Comment (AffectedArc07) [inline SQL/updates/60-61.py:43]: I assume this means the script will run from anywhere, like if the CWD is the repo root, it will load the data properly?

**Comment by AffectedArc07** [comment]
To make CI work, append `tools/ci/generate_sql_scripts.py`

```py
# And write the files and tell them to be used
for file in orderedSqlFiles:
    if file.endswith(".py"):
        # Begin snowflakery
        if file == "16-17.py":
            scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb feedback round\n")
        elif file == "17-18.py":
            scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb feedback feedback_2\n")
        elif file == "31-32.py":
            scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb\n")
        elif file == "38-39.py":
            scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb\n")
        else:
            print("ERROR: CI failed due to invalid python file in SQL/updates")
            exit(1)
```

And dont kill me for my python use, I wrote this script 4 years ago.

File: SQL/updates/60-61.py (hunk 1 above)
Comment (warriorstar-orion) [reply SQL/updates/60-61.py:43]: Yup:
```sh
$ echo "print(__file__)" >> tools/test.py
$ python tools/test.py
D:\ExternalRepos\third_party\Paradise\tools\test.py
```

**Comment by warriorstar-orion** [comment]
Will take a look at the players and saved_json_datums tables tomorrow.

**Comment by warriorstar-orion** [comment]
> To make CI work, append `tools/ci/generate_sql_scripts.py`
> 
> ```python
> # And write the files and tell them to be used
> for file in orderedSqlFiles:
>     if file.endswith(".py"):
>         # Begin snowflakery
>         if file == "16-17.py":
>             scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb feedback round\n")
>         elif file == "17-18.py":
>             scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb feedback feedback_2\n")
>         elif file == "31-32.py":
>             scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb\n")
>         elif file == "38-39.py":
>             scriptLines.append("python3 SQL/updates/" + str(file) + " 127.0.0.1 root root paradise_gamedb\n")
>         else:
>             print("ERROR: CI failed due to invalid python file in SQL/updates")
>             exit(1)
> ```
> 
> And dont kill me for my python use, I wrote this script 4 years ago.

Done.

**Comment by warriorstar-orion** [comment]
@AffectedArc07 I added stanzas to the migration to update the `characters`.`gear` column and `json_datum_saves`.`slotjson` column. Any other important JSON columns I'm missing?

File: config/example/config.toml [hunk 2]
```diff
@@ -181,7 +181,7 @@ ipc_screens = [
 # Enable/disable the database on a whole
 sql_enabled = false
 # SQL version. If this is a mismatch, round start will be delayed
-sql_version = 61
+sql_version = 62
```
Comment (Burzah) [inline config/example/config.toml:184]:
```suggestion
sql_version = 63
```

**Review by Burzah — CHANGES_REQUESTED** [review]
SQL version bump!
Make sure to change the script from `61-62.py` to `62-63.py`

File: config/example/config.toml (hunk 2 above)
Comment (warriorstar-orion) [reply config/example/config.toml]: Done

**Review by Burzah — APPROVED** [review]

**Review by AffectedArc07 — APPROVED** [review]

**Pull request merged**